### PR TITLE
Show swing ranges when making auxiliary or reserve choices

### LINKED ISSUE #690: Reserve die screen should mention swing ranges
When you're asked to add a reserve die, the game should remind you of your swing ranges (if you also have swing dice), since that might influence which die you choose.

### PR DESCRIPTION
**Pull request by blackshadowshade**
Fixes #690.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/730/

### CODE REVIEW

**Comment by cgolubi1** [comment]
Okay, what do we have here?
* I'm a little uncertain about the Game.js code to output the range itself --- is it a copy/paste of the code which outputs that stuff during swing die selection, or a reimplementation?  I'll take a look at that more carefully and have an opinion
* I'll try this out using some reserve/aux swing games, and form an opinion about it
* If i understand the code correctly, there should be no changes to the API results for any games *unless* those games have aux or reserve dice in them.  I'll test that theory by playing back some games.
* I'll play back some games with aux and reserve, and verify that the API changes are as expected and can be trivially fixed
* I'll play some new games to look for internal errors

**Comment by cgolubi1** [comment]
Okay, so, here's what i see.

First, just as a point of comparison --- this is what the swing die selection screen looks like --- it hasn't changed.  When you're choosing your own swing dice, the column on the right shows an informational display containing the valid ranges for your opponent's swing dice.  They're shown in the format:
* One swing die type per line
* style looks like: "X: (4-20)"

Here's the visual on that, and the reason i'm drawing attention to this is that IMO the right thing to do is to parrot that look when doing exactly the same thing during other game phases:

informational_swing_ranges_during_swing_selection

**Comment by cgolubi1** [comment]
Okay, now, with that in mind, here's the visual i see during reserve selection - the ranges are shown, but the style is totally different (all swing types on one line, comma-separated, no parens):

choose_reserve_shows_ranges

**Comment by cgolubi1** [comment]
And here's what i see during choose auxiliary: i looked in the API JSON, and the swing ranges *are* passed to the UI here, but nothing is shown in the UI:

choose_aux_no_ranges

**Comment by blackshadowshade** [comment]
So, is it just the missing parentheses in the reserve screen that are missing, or do you think that the layout should also be different?

**Comment by cgolubi1** [comment]
> So, is it just the parentheses in the reserve screen that are missing, or do you think that the layout should also be different?

There are two differences between the reserve UI and the "your opponent's swing ranges" part of the swing UI:
* the missing parentheses
* the fact that in the reserve UI the swing values are all on the same line, comma-separated, whereas in the swing UI, each value is on its own line

I think the reserve case should look like the swing case in terms of both of those points.

**Comment by cgolubi1** [comment]
One implementation suggestion: the code which adds the possible swing values in the SpecifyDice case is actually in `Game.actionSpecifyDiceActive()`, not in `Game.dieRecipeTable()` --- maybe that code should be split into its own function, and used by each of `Game.actionSpecifyDiceActive()`, `Game.actionChooseReserveDiceActive()`, and `Game.actionChooseAuxiliaryDiceActive()`.

**Comment by cgolubi1** [comment]
Hmm, also, something's not right with the backend for the auxiliary stuff --- if i create a game of kaufman (who has no swing dice) vs. Montague(u) (who has a swing die and an auxiliary swing die), the JSON contains:

```
"swingRequestArray":[],                                 # in playerDataArray[0]
"swingRequestArray":{"X":[4,20],"V":[6,12]},  # in playerDataArray[1]
```

That can't be right --- even if you fix the UI, the kaufman player won't see the information about what "V" values are possible.

**Comment by cgolubi1** [comment]
Yeah, this is consistent --- if one player has a swing auxiliary die and the other player has no auxiliary die, the player without the aux needs to be given that swing value in his swingRequestArray (assuming the implementation as you've done it in this pull), and is not.  Any chance an existing responderTest shows this behavior?  I'm checking now.

**Comment by cgolubi1** [comment]
Yes, indeed --- `test_interface_game_009()` (Merlin vs. Crane) --- Crane is offered a `+(X)`, so for your solution to work, Crane needs to have a swingRequestArray containing `X`, and doesn't.

**Comment by cgolubi1** [comment]
AFAICT, the API stuff for reserve looks okay.  So, summary of issues:
* If one player has a swing auxiliary and the other player has no auxiliary, the no-aux player gets no swingRequestArray, as in the case of `test_interface_game_009()`
* Display of swing values during CHOOSE_AUXILIARY appears not to work
* Display of swing values during CHOOSE_RESERVE looks different from, and is implemented inconsistently to, display of swing values during SPECIFY_DICE

**Comment by blackshadowshade** [comment]
This should now be working, tested with:
- Jedite vs Lancelot
- Zomulgastar vs Merlin

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/795/ (if it passes)

**Comment by blackshadowshade** [comment]
So, you'll notice that I've added a stub for a QUnit test. I'll need a hand getting an appropriate test written here.

**Comment by blackshadowshade** [comment]
http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/800/

**Comment by blackshadowshade** [comment]
http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/807/ (if it passes)

**Comment by cgolubi1** [comment]
Okay, where are we?
* Look at the changes shadowshade made to Game.js and make sure they look reasonable
* Look at the new backend changes to add the swing info for auxiliary dice and make sure they look reasonable
* Look at the new UI and make sure i like it visually
* Propose some contents for the test_Game.swingRangeTable test
* Play back games without aux or reserve again, and make sure there are still no changes
* Play back some games with aux and reserve, and verify that the API changes are as expected and can be trivially fixed
* Play some new RandomAI games and look for internal errors

**Comment by cgolubi1** [comment]
Okay, visually, something isn't right here --- there are two things, one is a clear bug, and one is a nit but one that i think we should get consistent or have a clear understanding of why we aren't.

The clear bug:
merlin_crane_double_x

In case it matters, you see the same behavior whether you look at the game as Crane or as Merlin:
crane_merlin_double_x

The nit is also visible in those same diffs --- sometimes the opponent's swing choice is displayed as 'X: (4-20)' (with a colon) and sometimes it's displayed as 'X (4-20)' (no colon).  My feeling is that the former way, with a colon, is better --- that's what the current site has now, and i think it's tidier looking:
live_site_swing_visuals

But mostly i want to insist that every time we display an "FYI" swing range (one where you're not typing a number into the box), we do it consistently.  If you feel strongly that there shouldn't be a colon, i'm fine with that, as long as you make it so that there's never one.

**Comment by cgolubi1** [comment]
I expect you'll make a bunch of changes finding and fixing the doubled-X, so i'm not going to look at this further, but instead throw it back over the fence.  Complain if that makes you sad, but i think it's the time-saving choice.

**Comment by blackshadowshade** [comment]
Okay, this logic should be fixed now. Tested with Merlin vs Rikachu.

If this is okay, then we can start working on adding unit tests to lock in this logic.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/821/ (if it passes)

**Comment by cgolubi1** [comment]
Alright, i threw this back over the fence after the doubled-X bug, which may have just been a JS bug after all.  So that means i still have a fair bit to do:
* Verify that the Merlin/Crane double-X bug is fixed
* Look at the changes shadowshade made to Game.js and make sure they look reasonable
* Look at the new backend changes to add the swing info for auxiliary dice and make sure they look reasonable
* Look at the new UI and make sure i like it visually
* Propose some contents for the test_Game.swingRangeTable test
* Play back games without aux or reserve again, and make sure there are still no changes
* Play back some games with aux and reserve, and verify that the API changes are as expected and can be trivially fixed
* Play some new RandomAI games and look for internal errors

**Comment by cgolubi1** [comment]
Okay, here's a sample test that passes for me.  In general, the point is that most functions are not truly modular, but take advantage of variables in the Api namespace which were set from the loadGameData response at initial page setup time.  So it doesn't make sense to test in a vacuum, but instead you pick some test game that has interesting data, call `Game.getCurrentGame()` to load that data, and pass it a function which contains your test.

That means that, for a function like your new one which takes arguments and returns HTML, once you've loaded one set of data, you can call your function with as many arguments as you'd like and test anything you'd like pertaining to the data you've loaded.  (My example has two tests, one without `allowInput` and one with.)  But if you also want to test how your function behaves on e.g. a game with reserve dice, you need a separate test function for that.  My naming convention is `test_Game.swingRangeTable` for some base test functionality (and that's what the regression tests will insist exists), and `test_Game.swingRangeTable_somedescription` to test some other game type of interest.

See if that helps.

```javascript
diff --git a/test/src/ui/js/BMTestUtils.js b/test/src/ui/js/BMTestUtils.js
index 6e33f32..2ec6f05 100644
--- a/test/src/ui/js/BMTestUtils.js
+++ b/test/src/ui/js/BMTestUtils.js
@@ -113,6 +113,7 @@ BMTestUtils.testGameId = function(gameDesc) {
   if (gameDesc == 'blackomega_thefool_reacttoinitiative') { return '2302'; }
   if (gameDesc == 'blackomega_thefool_captured_value_die') { return '2306'; }
 
+  if (gameDesc == 'merlin_ein_reacttoauxiliary_player') { return '2401'; }
   if (gameDesc == 'merlin_ein_reacttoauxiliary_nonplayer') { return '2402'; }
   if (gameDesc == 'merlin_ein_reacttoauxiliary_inactive') { return '2403'; }
 
diff --git a/test/src/ui/js/test_Game.js b/test/src/ui/js/test_Game.js
index 799bb89..58f95f7 100644
--- a/test/src/ui/js/test_Game.js
+++ b/test/src/ui/js/test_Game.js
@@ -1250,7 +1250,28 @@ test("test_Game.dieRecipeTable_chance", function(assert) {
 });
 
 test("test_Game.swingRangeTable", function(assert) {
-// james: this is currently empty, pending an appropriate test
+  stop();
+  BMTestUtils.GameType = 'merlin_ein_reacttoauxiliary_player';
+  Game.getCurrentGame(function() {
+
+    // First test without allowInput or showPrev
+    var htmlobj = Game.swingRangeTable(
+      Api.game.player.swingRequestArray, 'swing_range_table', false, false
+    );
+    var html = $('<div>').append(htmlobj.clone()).remove().html();
+    assert.deepEqual(html, '<table id="swing_range_table"><tbody><tr><td>X: (4-20)</td></tr><tr><td>Y: (1-20)</td></tr></tbody></table>',
+      "Swing range table has the expected contents without allowInput or showPrev");
+
+    // Second test with allowInput
+    var htmlobj = Game.swingRangeTable(
+      Api.game.player.swingRequestArray, 'swing_range_table', true, false
+    );
+    var html = $('<div>').append(htmlobj.clone()).remove().html();
+    assert.deepEqual(html, '<table id="swing_range_table"><tbody><tr><td>X (4-20):</td><td><input maxlength="2" id="swing_X" class="swing" type="text"></td></tr><tr><td>Y (1-20):</td><td><input maxlength="2" id="swing_Y" class="swing" type="text"></td></tr></tbody></table>',
+      "Swing range table has the expected contents with allowInput and not showPrev");
+
+    start();
+  });
 });
 
 test("test_Game.dieTableEntry", function(assert) {
```

**Comment by blackshadowshade** [comment]
Okay, this is odd. When I add that test (after merging master), it fails, since the Y line is missing. I presume that the expected is actually correct, so what's going on here?


QUnit output:

1. Swing range table has the expected contents without allowInput or showPrev
Expected:	
"<table id=\"swing_range_table\"><tbody><tr><td>X: (4-20)</td></tr><tr><td>Y: (1-20)</td></tr></tbody></table>"
Result:	
"<table id=\"swing_range_table\"><tbody><tr><td>X: (4-20)</td></tr></tbody></table>"
Diff:	
 "<table  id=\"swing_range_table\"><tbody><tr><td>X: (4-20)</td></tr><tr><td>Y: (1-20)</td></tr></tbody></table>" (4-20)</td></tr></tbody></table>" 
Source:	
http://localhost:8178/test-ui/js/test_Game.js:1262:21
http://localhost:8178/ui/js/Env.js:543:22
http://localhost:8178/ui/js/Api.js:80:28
f@http://localhost:8178/ui/js/extern/jquery-1.9.0.min.js:1:18379
fireWith@http://localhost:8178/ui/js/extern/jquery-1.9.0.min.js:1:37270
r@http://localhost:8178/ui/js/extern/jquery-1.9.0.min.js:3:16020
r@http://localhost:8178/ui/js/extern/jquery-1.9.0.min.js:3:21639
Swing range table has the expected contents with allowInput and not showPrev
Expected:	
"<table id=\"swing_range_table\"><tbody><tr><td>X (4-20):</td><td><input maxlength=\"2\" id=\"swing_X\" class=\"swing\" type=\"text\"></td></tr><tr><td>Y (1-20):</td><td><input maxlength=\"2\" id=\"swing_Y\" class=\"swing\" type=\"text\"></td></tr></tbody></table>"
Result:	
"<table id=\"swing_range_table\"><tbody><tr><td>X (4-20):</td><td><input type=\"text\" class=\"swing\" id=\"swing_X\" maxlength=\"2\"></td></tr></tbody></table>"
Diff:	
 "<table  id=\"swing_range_table\"><tbody><tr><td>X  (4-20):</td><td><input maxlength=\"2\" type=\"text\" class=\"swing\"  id=\"swing_X\" class=\"swing\" type=\"text\"></td></tr><tr><td>Y (1-20):</td><td><input maxlength=\"2\" id=\"swing_Y\" class=\"swing\" type=\"text\"></td></tr></tbody></table>" maxlength=\"2\"></td></tr></tbody></table>" 
Source:	
http://localhost:8178/test-ui/js/test_Game.js:1270:21
http://localhost:8178/ui/js/Env.js:543:22
http://localhost:8178/ui/js/Api.js:80:28
f@http://localhost:8178/ui/js/extern/jquery-1.9.0.min.js:1:18379
fireWith@http://localhost:8178/ui/js/extern/jquery-1.9.0.min.js:1:37270
r@http://localhost:8178/ui/js/extern/jquery-1.9.0.min.js:3:16020
r@http://localhost:8178/ui/js/extern/jquery-1.9.0.min.js:3:21639

**Comment by cgolubi1** [comment]
Hmm... so i did something mildly stupid in the example test --- i copy/pasted the definitions of htmlobj and html, including the `var `.  That's not good practice, and, while my QUnit didn't mind it, i wonder if yours did, and therefore ignored the second set of definitions somehow and ran the test against the first value of html.

**Comment by blackshadowshade** [comment]
Okay, I made a small change to responderTest, which gets the QUnit test working correctly now in Firefox. However, in Safari, we now have the failure that you were expecting, namely that the order of attributes is brittle.

Expected:	
"<table id=\"swing_range_table\"><tbody><tr><td>X (4-20):</td><td><input maxlength=\"2\" id=\"swing_X\" class=\"swing\" type=\"text\"></td></tr><tr><td>Y (1-20):</td><td><input maxlength=\"2\" id=\"swing_Y\" class=\"swing\" type=\"text\"></td></tr></tbody></table>"
Result:	
"<table id=\"swing_range_table\"><tbody><tr><td>X (4-20):</td><td><input type=\"text\" class=\"swing\" id=\"swing_X\" maxlength=\"2\"></td></tr><tr><td>Y (1-20):</td><td><input type=\"text\" class=\"swing\" id=\"swing_Y\" maxlength=\"2\"></td></tr></tbody></table>"
Diff:	
 "<table  id=\"swing_range_table\"><tbody><tr><td>X  (4-20):</td><td><input maxlength=\"2\" type=\"text\" class=\"swing\"  id=\"swing_X\" class=\"swing\" type=\"text\"></td></tr><tr><td>Y maxlength=\"2\"></td></tr><tr><td>Y  (1-20):</td><td><input maxlength=\"2\" type=\"text\" class=\"swing\"  id=\"swing_Y\" class=\"swing\" type=\"text\"></td></tr></tbody></table>" maxlength=\"2\"></td></tr></tbody></table>" 
Source:	
http://localhost:8178/test-ui/js/test_Game.js:1271:21
http://localhost:8178/ui/js/Env.js:543:22
http://localhost:8178/ui/js/Api.js:80:28
f@http://localhost:8178/ui/js/extern/jquery-1.9.0.min.js:1:18379
fireWith@http://localhost:8178/ui/js/extern/jquery-1.9.0.min.js:1:37270
r@http://localhost:8178/ui/js/extern/jquery-1.9.0.min.js:3:16020
r@http://localhost:8178/ui/js/extern/jquery-1.9.0.min.js:3:21639


Could you give me an example of how to check this structure component by component?

**Comment by cgolubi1** [comment]
Yeah, so if you look at something like `test_Game.dieBorderTogglePlayerHandler()`, that has an example which is exactly this sort of thing:
* You instantiate an object from the page (as you're already doing)
* You use that append/clone/remove hack to get the full HTML as a string (as you're already doing)
* You use `assert.ok(html.match(somestring), "some text");` to check that the HTML contains some contents

You can also use `assert.ok(!html.match(somestring), "some text");` to check that the HTML doesn't contain other contents.

**Comment by blackshadowshade** [comment]
Right, I think that I've made that QUnit test less brittle.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/850/ (if it passes)

**Comment by cgolubi1** [comment]
Looks good to me!  From my original test plan, the only things left to do here are the API regression tests with and without auxiliary and reserve.

File: test/src/api/responderTest.php [hunk 1]
```diff
@@ -4767,6 +4768,8 @@ public function test_interface_game_009() {
         // now load the game and check its state
         $retval = $this->verify_api_loadGameData($expData, $gameId, 10);
 
+        // the courtesy die is no longer offered after this
+        $expData['playerDataArray'][1]['swingRequestArray'] = array();
```
Comment (cgolubi1) [inline test/src/api/responderTest.php]: Okay, so, a side point here: the flow of logic in responderTest should be:

1. Take an action and make sure it makes sense
2. Modify `$expData` to match the new state of the world after the action
3. use `verify_api_loadGameData()` to test `$expData`

Here you're modifying `$expData` (item 2) *before* taking the action (item 1), which makes it unnecessarily difficult to see what's going on.

**Comment by cgolubi1** [comment]
So, i included that nit about the order of changes, because, in fact, you're introducing a regression in test game 009.

Here's what the game looks like from responder003's POV before responder004 decides whether to use auxiliary dice:

auxiliary_swing_before_opponent_choice

And here's what the game looks like from responder003's POV after responder004 decides whether to use auxiliary dice:

auxiliary_swing_after_opponent_choice

So, that's no good, and it's what the logic in the "Move 1" section of the game_009 test is supposed to catch.

Let me know if that doesn't make sense (either the bug or the comment about how the API regression stuff is supposed to work).

BTW, in looking at this, i notice that we may have introduced another regression in an earlier pull, and didn't catch it --- on line 4828 you can see that you set a property on the auxiliary die.  Looking at this code, i think there's probably a tidier way to check for these regressions in responder003's view while checking for the expected changes in responder004's view, and i can introduce that at some point.  The `$expData004` thing is a red herring --- PHP doesn't deepcopy, so that's never going to do what you want.

But that's really a side note --- please don't spend a lot of time worrying about deepcopy on that account.  For this pull, we should worry about the regression this pull is introducing, and it's more important than the previous one because it causes a UI-visible change on opponent's auxiliary choice, and most players are using the UI.

**Comment by cgolubi1** [comment]
Okay, here you go --- this is my proposed refactor of test 009.  The idea is just to make it jump out more when you're introducing a regression.  Adopt it if you like it:
```diff
diff --git a/test/src/api/responderTest.php b/test/src/api/responderTest.php
index f61e457..7cfc448 100644
--- a/test/src/api/responderTest.php
+++ b/test/src/api/responderTest.php
@@ -4759,6 +4759,9 @@ class responderTest extends PHPUnit_Framework_TestCase {
      * 14. responder004 gained initiative by turning down focus dice: f(6) from 4 to 1, f(8) from 8 to 1
      * 15. responder004 performed Skill attack using [(10):1,(12):8] against [s(20):9]; Defender s(20) was captured; Attacker (10) rerolled 1 => 2; Attacker (12) rerolled 8 => 7
      */
+    /**
+     * @group fulltest_deps
+     */
     public function test_interface_game_009() {
 
         // responder003 is the POV player, so if you need to fake
@@ -4775,35 +4778,44 @@ class responderTest extends PHPUnit_Framework_TestCase {
             array(),
             'responder003', 'responder004', 'Merlin', 'Crane', 3);
 
-        // Initial expected game data object
-        $expData = $this->generate_init_expected_data_array($gameId, 'responder003', 'responder004', 3, 'CHOOSE_AUXILIARY_DICE');
-        $expData['gameSkillsInfo'] = $this->get_skill_info(array('Auxiliary', 'Focus', 'Shadow'));
-        $expData['playerDataArray'][0]['swingRequestArray'] = array('X' => array(4, 20));
-        $expData['playerDataArray'][1]['swingRequestArray'] = array('X' => array(4, 20));
-        $expData['playerDataArray'][0]['button'] = array('name' => 'Merlin', 'recipe' => '(2) (4) s(10) s(20) (X) +s(X)', 'artFilename' => 'merlin.png');
-        $expData['playerDataArray'][1]['button'] = array('name' => 'Crane', 'recipe' => '(4) f(6) f(8) (10) (12)', 'artFilename' => 'crane.png');
-        $expData['playerDataArray'][0]['activeDieArray'] = array(
-            array('value' => NULL, 'sides' => 2, 'skills' => array(), 'properties' => array(), 'recipe' => '(2)', 'description' => '2-sided die'),
-            array('value' => NULL, 'sides' => 4, 'skills' => array(), 'properties' => array(), 'recipe' => '(4)', 'description' => '4-sided die'),
-            array('value' => NULL, 'sides' => 10, 'skills' => array('Shadow'), 'properties' => array(), 'recipe' => 's(10)', 'description' => 'Shadow 10-sided die'),
-            array('value' => NULL, 'sides' => 20, 'skills' => array('Shadow'), 'properties' => array(), 'recipe' => 's(20)', 'description' => 'Shadow 20-sided die'),
-            array('value' => NULL, 'sides' => NULL, 'skills' => array(), 'properties' => array(), 'recipe' => '(X)', 'description' => 'X Swing Die'),
-            array('value' => NULL, 'sides' => NULL, 'skills' => array('Auxiliary', 'Shadow'), 'properties' => array(), 'recipe' => '+s(X)', 'description' => 'Auxiliary Shadow X Swing Die'),
-        );
-        $expData['playerDataArray'][1]['activeDieArray'] = array(
-            array('value' => NULL, 'sides' => 4, 'skills' => array(), 'properties' => array(), 'recipe' => '(4)', 'description' => '4-sided die'),
-            array('value' => NULL, 'sides' => 6, 'skills' => array('Focus'), 'properties' => array(), 'recipe' => 'f(6)', 'description' => 'Focus 6-sided die'),
-            array('value' => NULL, 'sides' => 8, 'skills' => array('Focus'), 'properties' => array(), 'recipe' => 'f(8)', 'description' => 'Focus 8-sided die'),
-            array('value' => NULL, 'sides' => 10, 'skills' => array(), 'properties' => array(), 'recipe' => '(10)', 'description' => '10-sided die'),
-            array('value' => NULL, 'sides' => 12, 'skills' => array(), 'properties' => array(), 'recipe' => '(12)', 'description' => '12-sided die'),
-            array('value' => NULL, 'sides' => NULL, 'skills' => array('Auxiliary', 'Shadow'), 'properties' => array(), 'recipe' => '+s(X)', 'description' => 'Auxiliary Shadow X Swing Die'),
-        );
+	// Initial expected game data objects - maintain two parallel objects until we're done
+        // testing for information leakage about responder004's auxiliary choice
+        $initialExpData = array(
+            $this->generate_init_expected_data_array($gameId, 'responder003', 'responder004', 3, 'CHOOSE_AUXILIARY_DICE'),
+            $this->generate_init_expected_data_array($gameId, 'responder003', 'responder004', 3, 'CHOOSE_AUXILIARY_DICE'),
+        );
+        foreach (array(0, 1) as $idx) {
+            $initialExpData[$idx]['gameSkillsInfo'] = $this->get_skill_info(array('Auxiliary', 'Focus', 'Shadow'));
+            $initialExpData[$idx]['playerDataArray'][0]['swingRequestArray'] = array('X' => array(4, 20));
+            $initialExpData[$idx]['playerDataArray'][1]['swingRequestArray'] = array('X' => array(4, 20));
+            $initialExpData[$idx]['playerDataArray'][0]['button'] = array('name' => 'Merlin', 'recipe' => '(2) (4) s(10) s(20) (X) +s(X)', 'artFilename' => 'merlin.png');
+            $initialExpData[$idx]['playerDataArray'][1]['button'] = array('name' => 'Crane', 'recipe' => '(4) f(6) f(8) (10) (12)', 'artFilename' => 'crane.png');
+            $initialExpData[$idx]['playerDataArray'][0]['activeDieArray'] = array(
+                array('value' => NULL, 'sides' => 2, 'skills' => array(), 'properties' => array(), 'recipe' => '(2)', 'description' => '2-sided die'),
+                array('value' => NULL, 'sides' => 4, 'skills' => array(), 'properties' => array(), 'recipe' => '(4)', 'description' => '4-sided die'),
+                array('value' => NULL, 'sides' => 10, 'skills' => array('Shadow'), 'properties' => array(), 'recipe' => 's(10)', 'description' => 'Shadow 10-sided die'),
+                array('value' => NULL, 'sides' => 20, 'skills' => array('Shadow'), 'properties' => array(), 'recipe' => 's(20)', 'description' => 'Shadow 20-sided die'),
+                array('value' => NULL, 'sides' => NULL, 'skills' => array(), 'properties' => array(), 'recipe' => '(X)', 'description' => 'X Swing Die'),
+                array('value' => NULL, 'sides' => NULL, 'skills' => array('Auxiliary', 'Shadow'), 'properties' => array(), 'recipe' => '+s(X)', 'description' => 'Auxiliary Shadow X Swing Die'),
+            );
+            $initialExpData[$idx]['playerDataArray'][1]['activeDieArray'] = array(
+                array('value' => NULL, 'sides' => 4, 'skills' => array(), 'properties' => array(), 'recipe' => '(4)', 'description' => '4-sided die'),
+                array('value' => NULL, 'sides' => 6, 'skills' => array('Focus'), 'properties' => array(), 'recipe' => 'f(6)', 'description' => 'Focus 6-sided die'),
+                array('value' => NULL, 'sides' => 8, 'skills' => array('Focus'), 'properties' => array(), 'recipe' => 'f(8)', 'description' => 'Focus 8-sided die'),
+                array('value' => NULL, 'sides' => 10, 'skills' => array(), 'properties' => array(), 'recipe' => '(10)', 'description' => '10-sided die'),
+                array('value' => NULL, 'sides' => 12, 'skills' => array(), 'properties' => array(), 'recipe' => '(12)', 'description' => '12-sided die'),
+                array('value' => NULL, 'sides' => NULL, 'skills' => array('Auxiliary', 'Shadow'), 'properties' => array(), 'recipe' => '+s(X)', 'description' => 'Auxiliary Shadow X Swing Die'),
+            );
+        };
 
-        // now load the game and check its state
-        $retval = $this->verify_api_loadGameData($expData, $gameId, 10);
+        // now load the game from responder003's POV and check its state
+        $retval = $this->verify_api_loadGameData($initialExpData[0], $gameId, 10);
 
-        // the courtesy die is no longer offered after this
-        $expData['playerDataArray'][1]['swingRequestArray'] = array();
+        //////////
+	// WARNING - this section tests what each of responder003 and responder004 sees after responder004
+        // decides to keep the auxiliary die.  If you make any modifications to $initialExpData[0] between
+        // this message and the "END WARNING" comment, you are probably introducing information leakage
+        //////////
 
         ////////////////////
         // Move 01 - responder004 chose to use auxiliary die +s(X) in this game
@@ -4814,30 +4826,35 @@ class responderTest extends PHPUnit_Framework_TestCase {
             'Chose to add auxiliary die',
             $gameId, 'add', 5);
 
-        $expData004 = $expData;
-        $expData004['currentPlayerIdx'] = 1;
-        $expData004['playerDataArray'][0]['playerColor'] = '#ddffdd';
-        $expData004['playerDataArray'][1]['playerColor'] = '#dd99dd';
+        $initialExpData[1]['currentPlayerIdx'] = 1;
+        $initialExpData[1]['playerDataArray'][0]['playerColor'] = '#ddffdd';
+        $initialExpData[1]['playerDataArray'][1]['playerColor'] = '#dd99dd';
 
         // the API must tell the truth about whether the active player has
         // responded to auxiliary
-        $expData004['playerDataArray'][1]['waitingOnAction'] = FALSE;
-        $expData004['playerDataArray'][1]['activeDieArray'][5]['properties'] =
+        $initialExpData[1]['playerDataArray'][1]['waitingOnAction'] = FALSE;
+        $initialExpData[1]['playerDataArray'][1]['activeDieArray'][5]['properties'] =
             array('AddAuxiliary');
-        $expData004['gameActionLogCount'] = 0;
 
-        $retval = $this->verify_api_loadGameData($expData004, $gameId, 10);
+        // the courtesy die is no longer offered after this
+        $initialExpData[1]['playerDataArray'][1]['swingRequestArray'] = array();
+
+        $retval = $this->verify_api_loadGameData($initialExpData[1], $gameId, 10);
 
         $_SESSION = $this->mock_test_user_login('responder003');
 
-        // the API should lie about whether another player has responded to auxiliary
-        // to avoid information leaks
-        $expData['playerDataArray'][1]['waitingOnAction'] = TRUE;
-        $expData['gameActionLogCount'] = 0;
+        // BUG: this is a change to responder003's view of the game
+        $initialExpData[0]['playerDataArray'][1]['swingRequestArray'] = array();
 
-        // now load the game and check its state
-        $retval = $this->verify_api_loadGameData($expData, $gameId, 10);
+        // The API should lie about whether another player has responded to auxiliary
+        // to avoid information leaks - load the game from responder003's point of view
+        // using the unmodified $initialExpData[0] to test this
+        $retval = $this->verify_api_loadGameData($initialExpData[0], $gameId, 10);
 
+        //////////
+        // END WARNING - the remainder of this test will use $expData and test only responder003's view
+        //////////
+        $expData = $initialExpData[0];
 
         ////////////////////
         // Move 02 - responder003 chose to use auxiliary die +s(X) in this game
```

**Comment by blackshadowshade** [comment]
Okay, I think this is fixed.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/854/ (if it passes)

**Comment by cgolubi1** [comment]
Okay, so, i think we're back to:
* Manually play some aux games and make sure the regression appears gone
* Play back games without aux or reserve again, and make sure there are still no changes
* Play back some games with aux and reserve, and verify that the API changes are as expected and can be trivially fixed

**Comment by blackshadowshade** [comment]
http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/864/ (if it passes)

**Comment by cgolubi1** [comment]
I don't see any further issues, so i'm going to call this good.

**Pull request merged**